src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferConstCharOverConstUnitStringAnalyzer.cs

```diff
@@ -0,0 +1,122 @@
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    /// <summary>
+    /// Test for single character strings passed in to String.Append
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class PreferConstCharOverConstUnitStringAnalyzer : DiagnosticAnalyzer
+    {
+        public const string RuleId = "CA1831";
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.PreferConstCharOverConstUnitStringInStringBuilderTitle), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.PreferConstCharOverConstUnitStringInStringBuilderMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.PreferConstCharOverConstUnitStringInStringBuilderDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        internal static DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                                      s_localizableTitle,
+                                                                                      s_localizableMessage,
+                                                                                      DiagnosticCategory.Performance,
+                                                                                      RuleLevel.IdeSuggestion,
+                                                                                      s_localizableDescription,
+                                                                                      isPortedFxCopRule: false,
+                                                                                      isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterSyntaxNodeAction(AnalyzeSymbol, SyntaxKind.InvocationExpression);
```

Comment:
It's better to register an invocation operation rather than a node analyzer and then access the symbol.

Reply:
Yes, this should be rewritten as an IOperation analyzer. Please ping me if you need assistance.

Reply:
Done. I've seen other PRs where the `OperationAnalysisContext` object has another `RegisterOperationAction` call on it which finally does the real work. Would someone please explain what the `IOperation` is preferred over working on the SyntaxNode directly? Is it a performance concern?

Reply:
Semantic model + syntax tree provides an equivalent semantic view as an IOperation tree (semantic tree). The later is a newer API with primary advantages in space of perf:

1. `Syntax Tree <--> Operation tree` have a many-to-many mapping between nodes, so querying via semantic model + syntax node to fetch semantics is much more expensive then using and navigating the operation tree.
2. Operation tree is generated directly from the bound nodes or the abstract syntax tree that compiler uses as part of compiling code. This means compiler and analyzers share the semantic analysis phase, versus the former approach where coming from syntax, the compiler needs to do additional work to compute requested semantic information.
3. Using IOperation API you can write language-agnostic semantic analyzers that work for all languages that implement the IOperation API (currently C# and VB, but theoretically any language can implement it). SyntaxNode analyzers are generally required when you are doing some kind of syntactic analysis. Most semantic analysis should be implementable with IOperation.